### PR TITLE
Shared Typed Array

### PR DESCRIPTION
**Pull request by smatias1989**
Adds support for SharedTypedArray.

### CODE REVIEW

**Comment by smatias1989** [comment]
@dschenkelman  promises.test.js's rejected promises tests are currently fixed on b46cbddb859b2ae987038db7f41157262b8528f4.
Should we merge this pull request?

**Comment by smatias1989** [comment]
@dschenkelman a522d165ddfa59c1d2aefc9f1b86acf281ce2933 adds _supported_array_types_helper.js_ which is used on not-shared-typed-arrays tests to support shared typed arrays on them if they are available on the target test browser. I think that we have covered the shared array type tests with it.

File: dist/p-j-s-standalone.js [hunk 1]
```diff
@@ -106,49 +106,49 @@ var finisher = {
...
-  this.operation = operation; //todo: (mati) por ahora lo dejo para finalizar el reduce correctamente
-  this.globalContext = globalContext;
-  this.chainContext = chainContext;
-  previousOperations = previousOperations || [];
-  previousOperations.push(operation);
+var Chain = function (options) {
+  this.packager = new JobPackager(options.parts, options.source);
+  this.source = options.source;
+  this.parts = options.parts;
+  this.operation = options.operation; //todo: (mati) por ahora lo dejo para finalizar el reduce correctamente
```
Comment (dschenkelman) [inline dist/p-j-s-standalone.js]: english

File: dist/p-j-s-standalone.js [hunk 2]
```diff
@@ -186,23 +186,40 @@ Chain.prototype.reduce = function (reducer, seed, identityReducer, identity, loc
...
     var packs = self.packager.generatePackages(self.operations, self.chainContext);
-
     workers.sendPacks(packs, function(err, results){
-      if (err) { if (done) { done(err); } reject(err); return; }
-      var partial_results = results.map(function(result){
-        return new TypedArrayConstructor(result.value).subarray(0, result.newLength);
-      });
-      var m = merge_typed_arrays(partial_results);
+      if (err) { if (done) { done(err); } else { reject(err); } return; }
+      var m;
```
Comment (dschenkelman) [inline dist/p-j-s-standalone.js]: better variable name

File: examples/sapiaPicture/default-pjs.js [hunk 3]
```diff
@@ -39,18 +39,19 @@ var run;
       var tempContext = canvas.getContext("2d");
 
       tempContext.drawImage(source, 0, 0, canvas.width, canvas.height);
-
       var canvasData = tempContext.getImageData(0, 0, canvas.width, canvas.height);
+      var temp = new SharedUint8ClampedArray(canvasData.data.length);
```
Comment (dschenkelman) [inline examples/sapiaPicture/default-pjs.js]: there should be a separate samples for shared array buffers and normal array buffers. just copy paste and change the array types.

File: examples/sapiaPicture/defaultnoworker.js [hunk 4]
```diff
@@ -44,8 +44,11 @@ var processSepia = function (binaryData, l) {
 
         var canvasData = tempContext.getImageData(0, 0, canvas.width, canvas.height);
         var start = new Date();
+        console.time('serial');
```
Comment (dschenkelman) [inline examples/sapiaPicture/defaultnoworker.js]: remove console.time calls

File: gulpfile.js [hunk 5]
```diff
@@ -33,9 +33,11 @@ gulp.task('build-browserify', function() {
     .pipe(browserified())
     .pipe(rename(projectName + '.js'))
     .pipe(gulp.dest('./dist/'))
+    .pipe(gulp.dest('./examples/sapiaPicture'))
```
Comment (dschenkelman) [inline gulpfile.js]: this is no longer necessary, was just to simplify the iteration process

File: src/chain.js [hunk 6]
```diff
@@ -110,15 +110,22 @@ function createChain(oldChain, options){
...
-    var TypedArrayConstructor = self.source.constructor;
     var packs = self.packager.generatePackages(self.operations, self.chainContext);
-
     workers.sendPacks(packs, function(err, results){
       if (err) { if (done) { done(err); } else { reject(err); } return; }
-      var partial_results = results.map(function(result){
-        return new TypedArrayConstructor(result.value).subarray(0, result.newLength);
-      });
-      var m = merge_typed_arrays(partial_results);
+      var m;
```
Comment (dschenkelman) [inline src/chain.js]: better variable name

File: src/chain.js [hunk 7]
```diff
@@ -130,4 +137,23 @@ Chain.prototype.__verifyPreviousOperation = function () {
   }
 };
 
+Chain.prototype.__shouldMergeSeparatedBuffers = function (typedArrayType, operations) {
+  if (typedArrayType.indexOf('Shared') > -1) {
```
Comment (dschenkelman) [inline src/chain.js]: We should check if it is `=== 0` meaning that it starts with that

File: src/chain.js [hunk 8]
```diff
@@ -130,4 +137,23 @@ Chain.prototype.__verifyPreviousOperation = function () {
...
+    return this.__arrayReducerOperationWasApplied(operations);
+  }
+  return true;
+};
+
+Chain.prototype.__arrayReducerOperationWasApplied = function (operations) {
```
Comment (dschenkelman) [inline src/chain.js]: instead of reducer (related to reduce) let's use: `__arrayChangingSizeOperationWasApplied`

File: src/utils.js [hunk 9]
```diff
@@ -49,6 +116,13 @@ utils.isTypedArray = function (obj) {
   }
 };
 
+utils.isSharedArray = function (obj) {
+  if (!obj) {
+    return false;
+  }
+  return utils.getTypedArrayType(obj).indexOf('Shared') > -1;
```
Comment (dschenkelman) [inline src/utils.js]: same as in other comment `=== 0`

File: test/job_packager.test.js [hunk 10]
```diff
@@ -1,174 +0,0 @@
-'use strict';
-
-describe('job packager', function(){
-  var JobPackager = require('../src/job_packager');
-  var errors = require('../src/errors');
```
Comment (dschenkelman) [inline test/job_packager.test.js:5]: is this ok? removing all of it?

File: test/local_context.test.js [hunk 11]
```diff
@@ -6,6 +6,7 @@ describe('local context tests', function(){
   var chromeHelper = require('./chrome_version_helper');
   var utils = require('../src/utils.js');
   var normalSourceArray = [1,2,3,5,13,16,32,63,64,129,255,500,1001,1023,1024];
+  var supportedArrays = require('./supported_array_types_helper')();
```
Comment (dschenkelman) [inline test/local_context.test.js:9]: nice!

File: test/promises.test.js [hunk 12]
```diff
@@ -96,44 +96,26 @@ describe('promises tests', function () {
...
         var mapper = function (e, ctx) { return ctx.aux(e); };
         var promise = pjs(sourceArray).map(mapper).seq();
-        promise.then(function (result) {
-          done('should fail');
-        }, function (err) {
-          expect(err.name).to.equal('WorkerError');
-          expect(err.message).to.equal('Uncaught TypeError: undefined is not a function');
-          done();
-        });
+        return promise.should.be.rejected;
```
Comment (dschenkelman) [inline test/promises.test.js]: without checking the error?

File: test/job_packager.test.js (hunk 10 above)
Comment (smatias1989) [inline test/job_packager.test.js:5]: I changed the name to job_packager.test_s_.js. Then I merged dev to shared, that file thrown some conflicts and I forget to delete it.
I merge the conflicting changes into the renamed file before I delete it.

**Pull request merged**